File: app/src/main/java/com/example/vamshi/homwayprojectchallenge/Screens/MainScreen/MainActivity.java
package com.example.vamshi.homwayprojectchallenge.Screens.MainScreen;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.util.Log;
import android.widget.Toast;

import com.example.vamshi.homwayprojectchallenge.Dagger.Components.DaggerMainActivityComponent;
import com.example.vamshi.homwayprojectchallenge.Model.Retrofit.QueryConstants;
import com.example.vamshi.homwayprojectchallenge.MyApp;
import com.example.vamshi.homwayprojectchallenge.PlacesEntity;
import com.example.vamshi.homwayprojectchallenge.R;
import com.example.vamshi.homwayprojectchallenge.Screens.MapScreen.MapActivity;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.ArrayList;

import javax.inject.Inject;

import butterknife.ButterKnife;


public class MainActivity extends AppCompatActivity
        implements SearchView.OnQueryTextListener {

    RecyclerView.Adapter adapter;

    ArrayList<PlacesEntity> places;

    @Inject
    MainPresenter mPresenter;

    //  @BindView(R.id.search)
    SearchView searchView;
    //   @BindView(R.id.recyclerview)
    RecyclerView recyclerView;
    FloatingActionButton fab;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getSavedInstanceState(savedInstanceState);

        injectDependencies();
        setContentView(R.layout.activity_main);
        searchView = findViewById(R.id.search);
        recyclerView = findViewById(R.id.recyclerview);

        fab = findViewById(R.id.fabmap);
        fab.setOnClickListener((v) -> fabClick());

        searchView.setOnQueryTextListener(this);
        searchView.setOnCloseListener( () ->  onClose() );

        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        adapter = new PlacesAdapter(places, this, mPresenter);
        recyclerView.setAdapter(adapter);

    }


    private void injectDependencies() {
        ButterKnife.bind(this);
        DaggerMainActivityComponent
                .builder()
                .appComponent(((MyApp) getApplication()).getAppComponent())
                .build()
                .Inject(this);
    }


    @Override
    public boolean onQueryTextSubmit(String query) {

        mPresenter.getPlaces(query);

        return true;
    }

    @Override
    public boolean onQueryTextChange(String newText) {

        return false;
    }


    public boolean onClose() {
        if (places != null) {
            places.clear();
        }
        adapter = new PlacesAdapter(places, this, mPresenter);
        recyclerView.setAdapter(adapter);
        return false;
    }

    //  @OnClick(R.id.fabmap)
    void fabClick() {

        Intent startMap = new Intent(this, MapActivity.class);
        startMap.putParcelableArrayListExtra(QueryConstants.Mainbundlekey, places);
        startActivity(startMap);
        Toast.makeText(this, "Map", Toast.LENGTH_SHORT).show();

    }

    void getSavedInstanceState(Bundle savedInstanceState) {
        if (savedInstanceState != null) {
            places = savedInstanceState.
                    getParcelableArrayList(QueryConstants.Mainbundlekey);
            Log.i("bundle saved", "onRestoreInstanceState: " +
                    places +
                    savedInstanceState.containsKey(QueryConstants.Mainbundlekey));
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        mPresenter.maStart();
        EventBus.getDefault().register(this);
    }


    @Override
    protected void onSaveInstanceState(Bundle outState) {
        outState.putParcelableArrayList(QueryConstants.Mainbundlekey, places);
        super.onSaveInstanceState(outState);
    }

    @Override
    protected void onStop() {
        super.onStop();
        mPresenter.maStop();
        EventBus.getDefault().unregister(this);
    }


    @Override
    protected void onDestroy() {
        super.onDestroy();


    }

    // hello this is for git testing part
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void changePlacesAdapter(ArrayList<PlacesEntity> placeEvent) {

        places = placeEvent;
        adapter = new PlacesAdapter(places, this, mPresenter);
        recyclerView.setAdapter(adapter);

        Log.i("CHANGE PLACE ADAPTER", "changePlacesAdapter: yes");
    }


    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);

        getSavedInstanceState(savedInstanceState);
    }
}

File: app/src/main/java/com/example/vamshi/Curious/Screens/MainScreen/MainActivity.kt
package com.example.vamshi.Curious.Screens.MainScreen

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationManager
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.SearchView
import android.util.Log
import android.widget.Toast
import com.example.vamshi.Curious.Dagger.Components.DaggerMainActivityComponent
import com.example.vamshi.Curious.Model.Retrofit.QueryConstants
import com.example.vamshi.Curious.MyApp
import com.example.vamshi.Curious.PlacesEntity
import com.example.vamshi.Curious.R
import com.example.vamshi.Curious.Screens.MapScreen.MapActivity
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationServices
import com.google.android.gms.maps.model.LatLng
import kotlinx.android.synthetic.main.activity_main.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import javax.inject.Inject
import kotlin.math.round
import kotlin.math.truncate


class MainActivity : AppCompatActivity(), SearchView.OnQueryTextListener {

    private var places: ArrayList<PlacesEntity> = ArrayList()
    private lateinit var fusedLocationClient: FusedLocationProviderClient

    @Inject
    lateinit var mPresenter: MainPresenter

    private var mlocation: String = "47.6,-122.3"


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        getSavedInstanceState(savedInstanceState)
        injectDependencies()
        if (checkPermission()) { findLocation() }
        setContentView(R.layout.activity_main)

        fab_map.setOnClickListener { fabClick() }
        search_view.setOnQueryTextListener(this)
        search_view.setOnCloseListener { places?.clear(); false }

        recycler_view.setHasFixedSize(true)
        recycler_view.layoutManager = LinearLayoutManager(this)
        recycler_view.adapter = PlacesAdapter(places, this, mPresenter)
    }

    private fun injectDependencies() {
        DaggerMainActivityComponent
                .builder()
                .appComponent((application as MyApp).appComponent)
                .build()
                .Inject(this)
    }

    override fun onQueryTextSubmit(query: String) = mPresenter.getPlaces(query)

    override fun onQueryTextChange(newText: String) = false

    private fun fabClick() {
        val startMap = Intent(this, MapActivity::class.java)
        startMap.putParcelableArrayListExtra(QueryConstants.Mainbundlekey, places)
        startActivity(startMap)
        Toast.makeText(this, "Map", Toast.LENGTH_SHORT).show()
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun changePlacesAdapter(placeEvent: ArrayList<PlacesEntity>) {
        places = placeEvent
        recycler_view.adapter = PlacesAdapter(places, this, mPresenter)

        Log.i("CHANGE PLACE ADAPTER", "changePlacesAdapter: yes")
    }


    @SuppressLint("MissingPermission")
    fun findLocation(): Boolean {
        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
        fusedLocationClient.lastLocation
                .addOnSuccessListener { location: Location? -> mPresenter.currentLocation(location) }
        return true
    }


    private fun getSavedInstanceState(savedInstanceState: Bundle?) {
        savedInstanceState?.apply {
            places = this.getParcelableArrayList(QueryConstants.Mainbundlekey)

            Log.i("bundle saved", "onRestoreInstanceState: " + places +
                    savedInstanceState?.containsKey(QueryConstants.Mainbundlekey))
        }
    }

    override fun onStart() {
        super.onStart()
        mPresenter.maStart()
        EventBus.getDefault().register(this)
    }


    override fun onStop() {
        super.onStop()
        mPresenter.maStop()
        EventBus.getDefault().unregister(this)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putParcelableArrayList(QueryConstants.Mainbundlekey, places)
        super.onSaveInstanceState(outState)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        getSavedInstanceState(savedInstanceState)
    }


    fun checkPermission(): Boolean {

        val permissions = arrayOf(Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.ACCESS_COARSE_LOCATION)
        var granted = PackageManager.PERMISSION_GRANTED

        fun isAccessFineLocationEnabled(): Boolean {
            return ContextCompat.checkSelfPermission(this, permissions.get(0)) == granted
        }

        fun isAccessCoarseLocationEnabled(): Boolean {
            return ContextCompat.checkSelfPermission(this, permissions.get(1)) == granted
        }

        if (isAccessCoarseLocationEnabled() && isAccessFineLocationEnabled()) {
            return true
        } else {
            ActivityCompat.requestPermissions(this as Activity, permissions, 1234)
        }

        return false

    }


    override fun onRequestPermissionsResult(requestCode: Int,
                                            permissions: Array<String>,
                                            grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        when (requestCode) {
            1234 -> if (grantResults[0] != PackageManager.PERMISSION_GRANTED || grantResults[1] != PackageManager.PERMISSION_GRANTED) {

                Log.i("#######", "getPermission:yes")

            } else {

                findLocation()
            }
        }
    }
}